cachelib/allocator/CacheAllocator-inl.h

```diff
@@ -2664,7 +2544,7 @@ void CacheAllocator<CacheTrait>::evictForSlabRelease(
     auto owningHandle =
         item.isChainedItem()
             ? evictChainedItemForSlabRelease(item.asChainedItem())
```

Comment:
> Would you be OK with doing this refactor for Slab Release in separate PR (after this one would be merged)?

Yes that is totally fine.

> One specific concern with the approach you suggested is: when should we call nvmCache->put  If we would create putToken before removing the item from AccessContainer and MMContainer and issued the actual put after item.isOnlyMoving() it will result in keeping they key in the inFlightPut map longer - is this OK?

Oh we should issue a put to NvmCache earlier. As soon as we have removed the RAM parent item from access-container. This is because from DRAM cahce's perspective, it is already invisible and it's expected that this item is on the queue to get inserted into NvmCache. (Similar to the existing logic in evictChainedItemForSlabRelease() )

If we delay the insertion to NvmCache until after "isOnlyMoving() == true", we actually become more at risk of seeing a concurrent Lookup to the same key. In which case, we will be forced to give up the insertion altogether. (Since the concurrent lookup have to return quickly to inform the caller a key exists or not, and in this case we don't have the item inserted in NvmCache yet and we will be forced to drop the insertion altogether). This is to avoid a scenario like:

```
(1) T1 -> Created PutToken (not yet enqueued or still in the queue)
(2) T2 -> GET -> Miss
(3) T1 -> Inserted into NvmCache
(4) T2 -> GEt -> Hit <=== Bad. This is read/write inconsistency. This should return a miss as well.
```

Today in our logic, we handle this correctly by returning a miss at (4), but it's not ideal. Ideally we want (1) and (3) to be as close as possible to avoid (2) getting in between. Increasing the time gap between (1) and (3) will make (2) more likely to get interleaved.